### DISCUSSION TITLE
Has someone gotten the Surface Pen to work on a Surface 7 Pro ?

### DISCUSSION THREAD

**Discussion by Miklom** [linux-surface/linux-surface#593] [Q&A] [answered]
Hi guys, so I just made the switch from Windows to Linux (Manjaro KDE) on my Surface Pro 7 and installed the linux-surface kernal.
The Touchscreen is now working, but the pen is not. Is the Pen supported, and I just messed up something during the installation, or is there no support for it yet ? 
I appreciate any help you guys can give me :)

**Comment by qzed** [comment, marked as answer]
The pen does currently not work on any gen7 device. See https://github.com/linux-surface/intel-precise-touch/issues/4 for more details. Specifically https://github.com/linux-surface/intel-precise-touch/issues/4#issuecomment-926583916 for an overview of the current state of touch.

**Comment by imjustanotherdude** [comment]
Hi everyone, I installed Ubuntu on my Surface Pro 7 and found some pretty cool writing app called Stylus Labs Write. Everything is working fine, thank to thous who made that possible! But sadly the Pen does not work and that is a big problem in my opinion. 
I found this post from 5.Oct.2021 so i wonder if there is a solution for the Pen? 
Thank you in advance!

**Reply by NP-chaonay** [reply]
1) reinstall driver from A
2) and reinstall iptsd from B

A: From [this github](https://github.com/linux-surface/intel-precise-touch)  not software repo (rebuild manually from github) 
B: From [this github](https://github.com/linux-surface/iptsd)  not software repo (using prebuilt from Github Action or rebuild manually from github)

feel free to ask anything

**Reply by imjustanotherdude** [reply]
Ok sorry, but I have no clue how to do that...

**Reply by imjustanotherdude** [reply]
"feel free to ask anything" , ... ok

**Reply by qzed** [reply]
For (A) see the ["Building (DKMS)"](https://github.com/linux-surface/intel-precise-touch#building-dkms) section in the readme (i.e., clone/download the repo then run what's specified there). For (B) you can download a pre-built package from https://github.com/linux-surface/iptsd/actions (select the latest and download the respective artifact for your distribution, then install that).

**Reply by imjustanotherdude** [reply]
So I downloaded this (debian-latest) under "Artifacts"->  [](https://github.com/linux-surface/iptsd/actions/runs/3264533755) and removed the old build and installed a the new one. 
Everything works fine.
The pen (Microsoft) does connect via BT but does not work. I think i did something wrong? 

Sorry i am relatively new to Ubuntu / Linux...

**Reply by qzed** [reply]
Did you install the IPTS driver (A)?

**Reply by imjustanotherdude** [reply]
Thank you for your Help! 
I followed the instructions "Building (DKMS)" and i guess it was successful (I did not get any errors).
But the touch screen works only 3 seconds, after that I have to get my mouse / keyboard.... The Pen still does not work.

I installed: 
1. the Surface Kernel as shown here:  https://github.com/linux-surface/linux-surface/wiki/Installation-and-Setup
2. the IPTS driver (i guess)

How do i check if its set up all right, do i miss something?

**Reply by NP-chaonay** [reply]
ok so I think there is something wrong about touching I may could not solve that

But for pen If I not wrong, you can use it without BT.

btw does iptsd installed?

**Comment by Jeremycouturier** [comment]
Hi !

I am using a surface pro 7 with fedora 38 and the pen does work. Running fedora on the surface pro 7 really does make a cool linux tablet and I'm really grateful toward the developers to make that happen.

Even though the pen works, there are still some missing functionalities.

-For example, while the pen often behaves like scrolling under windows, it instead behaves like a regular mouse pointer under linux, which prevents using the stylus for scrolling content. It feels to me like solving this issue would really contribute to make the surface pro 7 even more tablet-like under linux.

-Another issue is the lag between the physical position of the pen tip and the cursor on the screen. On Windows 11, when crossing the screen with the pen, there is a ~2mm lag between the two. On Fedora, the lag is rather a good 3 to 4 cm instead. That makes it unpractical to use the pen for note-taking on linux using applications such as xournal++ because the writing feeling is pretty bad.

This last issue is the only reason why i keep a windows partition on my surface pro 7, so I can also boot into Windows when I want to use my surface pro 7 for note-taking.

But 95 % of the time I use the surface pro 7 for either work or entertainment, under the mighty Fedora. Thanks again to the developers.

Do the developers believe these pen issues will get solved in the future releases of the kernel ?

Jérémy

**Reply by StollD** [reply]
> Do the developers believe these pen issues will get solved in the future releases of the kernel ?

The pen coordinates are reported in the standard linux way, just like all other pens. Scrolling with the pen has to be supported by the desktop environment. I believe it already works in KDE / Qt apps.

I have experimented with letting the pen identify as a touchscreen to make scrolling with it work everywhere, but the downside of this is that drawing apps won't work anymore in that case.

Regarding the latency of the pen, this is a more difficult issue, with many layers to it.

* The linux driver for IPTS has to do polling because of how the hardware was designed around being used together with the GPU. A microcontroller in the GPU basically polls all the time, and schedules processing on the GPU when data arrives. On the main CPU we can't poll all the time, so we need to accept some latency. This is actually something that could be handled differently on these gen7 devices, e.g. SP7, SB3, SL3, but it is hard to get the behaviour right without a device to test on or at least a reference implementation, so for now they share the same behaviour as the older IPTS devices.
* Before the inputs arrive at an application, the data is sent to userspace, processed there, and then sent back to kernelspace through uinput. Every switch between kernel- and userspace takes time and adds latency. With the proprietary driver, the data never leaves kernelspace before inputs are sent to the applications.
* Applications themselves can have a certain latency, because of how they render or receive inputs.

Another important point is that the hardware itself has latency. Windows tries to hide it by using predictive algorithms, but this AFAIK is only done in some apps (e.g. OneNote). You can find reports on e.g. Reddit where people report that they have similar latency in Photoshop, but no latency in OneNote.

For more information, check this issue: https://github.com/linux-surface/linux-surface/issues/1061

**Reply by Jeremycouturier** [reply]
> I have experimented with letting the pen identify as a touchscreen to make scrolling with it work everywhere, but the downside of this is that drawing apps won't work anymore in that case.

Actually, since I am not using Fedora for drawing, I am very interested in letting the pen identify as a touchscreen, in order to be able to scroll instead of just selecting content. How do I do that ?

>I believe it already works in KDE

So switching from Gnome to KDE would allow for pen scrolling ? Anyway the above solution sounds better because I prefer gnome to KDE, but I don't really understand what you mean by letting the pen identify as a touchscreen?

**Reply by StollD** [reply]
> So switching from Gnome to KDE would allow for pen scrolling ?

In some applications it works, yes. But I don't know the exact criteria. I have only seen it in the settings app, and I doubt you want to spend much time there. Maybe its all QML apps?

> Actually, since I am not using Fedora for drawing, I am very interested in letting the pen identify as a touchscreen, in order to be able to scroll instead of just selecting content. How do I do that ?

You would have to modify iptsd. Its been a while since I tried it out, so I don't really know anymore what exactly you have to modify. I think I did something like removing all `BTN_TOOL_PEN`, `BTN_TOOL_RUBBER` and `BTN_STYLUS` events.

If you know C++, this is the file you would have to modify: https://github.com/linux-surface/iptsd/blob/master/src/apps/daemon/stylus.hpp.

**Reply by KTibow** [reply]
I can't get the pen to work, and I'm also on Fedora with linux-surface on a SP7. Do you need to compile something in order to make it work?

**Reply by StollD** [reply]
Please provide the output of `systemctl status $(iptsd-find-service)` and `dmesg`.

**Reply by KTibow** [reply]
```
● iptsd@dev-hidraw1.service - Intel Precise Touch & Stylus Daemon
     Loaded: loaded (/usr/lib/systemd/system/iptsd@.service; static)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: active (running) since Sun 2023-06-18 05:17:28 PDT; 7h ago
       Docs: https://github.com/linux-surface/iptsd
   Main PID: 814 (iptsd)
      Tasks: 1 (limit: 18631)
     Memory: 1.0M
        CPU: 596ms
     CGroup: /system.slice/system-iptsd.slice/iptsd@dev-hidraw1.service
             └─814 /usr/bin/iptsd /dev/hidraw1

Jun 18 05:17:28 surface-pro-7 systemd[1]: Started iptsd@dev-hidraw1.service - Intel Precise Touch & Stylus Daemon.
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.182] [info] Metadata:
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.182] [info] rows=44, columns=64
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.182] [info] width=25978, height=17319
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.182] [info] transform=[-412.3492,0,25978,0,-402.76746,17319]
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.182] [info] unknown=1, [178,182,180,1,178,182,180,1,90,171,100,20,172,177,175,2]
Jun 18 05:17:28 surface-pro-7 iptsd[814]: [05:17:28.194] [info] Connected to device 045E:099F
```
[dmesg.txt](https://github.com/linux-surface/linux-surface/files/11782464/dmesg.txt)

**Reply by StollD** [reply]
That looks like everything is working properly. What happens if you try to use the pen? Nothing at all, or does the pen register but misbehave?

**Reply by KTibow** [reply]
Nothing happens at all. Could be that my tip is in such a bad shape it doesn't register, but I doubt it. Touch works fine.

**Reply by qzed** [reply]
Maybe it's the pen battery?

**Reply by KTibow** [reply]
That was it. The battery was at like 0.3V. I'll see if it works with a replacement.

**Reply by KTibow** [reply]
The pen works! :tada:

**Reply by imjustanotherdude** [reply]
Hello everyone, I just installed Ubuntu 22.10 on my Pro 7 and everything works now:) even the pen! But I realized that there is a little input lag in note taking apps. I figured out that Rnote works the best, but on Windows it's just faster... It's not extrem, but it feels weird. Has someone experienced the same?

**Reply by StollD** [reply]
Yes, this is a common issue. There is some information and research in https://github.com/linux-surface/linux-surface/issues/1061. The TL;DR is that the hardware has a certain latency, and how we have to route and process the data that comes from the hardware under linux doesnt really help either.

It is also likely that Windows has to deal with similar issues and tries to hide the latency through input prediction, which has to be implemented in the application (or a common API designed for this, like Windows Ink).
